exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/DrillRestServer.java

```diff
@@ -204,9 +217,15 @@ public WebUserConnection provide() {
                 config.getLong(ExecConstants.HTTP_SESSION_MEMORY_RESERVATION),
                 config.getLong(ExecConstants.HTTP_SESSION_MEMORY_MAXIMUM));
 
+        // Create a dummy close future which is needed by Foreman only. Foreman uses this future to add a close
+        // listener to known about channel close event from underlying layer. We use this future to notify Foreman
+        // listeners when the Web connection between Web Client and WebServer is closed. This will help Foreman to cancel
+        // all the running queries for this Web Client.
```

Comment:
Classic Web communication is connectionless and stateless. For performance, some browsers use a keep-alive mechanism. But, this must be entirely optional. In this way, an HTTP connection is completely unlike an RPC connection.

Given that the HTTP connection will be immediately closed after submitting a query, do we really want to kill the query in response?

Or, are we using a web session concept that "closes" in response to a session timeout?

Reply:
The query is killed when the HttpSession is invalidated either by logout or by timeout in case of authenticated user session. But for unauthenticated or anonymous user we don't kill the query after HttpSession timeout since we don't maintain any session related information. The query will keep running unless it is explicitly cancelled from /profile page.